File: src/main/leetcode/p352/SummaryRanges.java
package p352;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Created by wangsenyuan on 5/31/16.
 */
public class SummaryRanges {
    /**
     * Initialize your data structure here.
     */
    private TreeMap<Integer, Integer> tree;

    public SummaryRanges() {
        tree = new TreeMap<>();
    }

    public void addNum(int val) {
        // Value existed as key
        if (tree.get(val) != null)
            return;

        // Value existed inside intervals
        Map.Entry<Integer, Integer> low = tree.lowerEntry(val);
        if (low != null && low.getValue() >= val)
            return;

        // Value can merge to end
        if (low != null && low.getValue() + 1 == val) {
            if (tree.get(val + 1) != null) {
                int start = low.getKey();
                int end = tree.get(val + 1);
                tree.remove(val + 1);
                tree.remove(low.getKey());
                tree.put(start, end);
            } else
                tree.put(low.getKey(), val);
            return;
        }

        // Value can merge to start
        Map.Entry<Integer, Integer> high = tree.higherEntry(val);
        if (high != null && high.getKey() == val + 1) {
            int end = high.getValue();
            tree.remove(high.getKey());
            tree.put(val, end);
            return;
        }

        // Isolated new value
        tree.put(val, val);
    }

    public List<Interval> getIntervals() {
        List<Interval> l = new LinkedList<>();

        for (Map.Entry<Integer, Integer> entry : tree.entrySet())
            l.add(new Interval(entry.getKey(), entry.getValue()));

        return l;
    }

}

File: src/main/leetcode/set300/set350/p352/SummaryRanges.scala
package set300.set350.p352


/**
  * Definition for an interval.
  **/
class Interval(var _start: Int = 0, var _end: Int = 0) {
  var start: Int = _start
  var end: Int = _end
}

class SummaryRanges() {

  /** Initialize your data structure here. */
  val tree = new java.util.TreeMap[Int, Int]()

  def addNum(v: Int) {
    if (!tree.containsKey(v)) {
      val low = tree.lowerEntry(v)

      if (low == null || low.getValue < v) {
        val high = tree.higherEntry(v)

        if (low != null && high != null && low.getValue + 1 == v && v + 1 == high.getKey) {
          //merge
          tree.remove(high.getKey)
          tree.put(low.getKey, high.getValue)
        } else if (low != null && low.getValue + 1 == v) {
          //extend
          tree.put(low.getKey, v)
        } else if (high != null && v + 1 == high.getKey) {
          tree.remove(high.getKey)
          tree.put(v, high.getValue)
        } else {
          tree.put(v, v)
        }
      }
    }
  }

  import scala.collection.mutable.ListBuffer


  def getIntervals(): List[Interval] = {
    val buf = ListBuffer.empty[Interval]
    val iter = tree.entrySet().iterator()
    while (iter.hasNext) {
      val entry = iter.next()
      val key = entry.getKey
      val value = entry.getValue
      buf += new Interval(key, value)
    }
    buf.toList
  }

}
